pkg/lib/infra/sms/custom/client.go

```diff
@@ -65,67 +76,101 @@ type SMSWebHook struct {
 	req, err := w.PrepareRequest(ctx, u, payload)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	resp, err := w.Client.Client.Do(req)
+	resp, err := w.Client.Do(req)
 	if err != nil {
-		return nil, err
+		return err
 	}
 	defer resp.Body.Close()
 
 	dumpedResponse, err := httputil.DumpResponse(resp, true)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
-		return dumpedResponse, nil
+	bodyBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return errors.Join(err, &smsapi.SendError{
+			DumpedResponse: dumpedResponse,
+		})
 	}
 
-	return nil, &smsapi.SendError{
-		DumpedResponse: dumpedResponse,
+	responseBody, err := ParseResponseBody(bodyBytes)
+	if err != nil {
+		// This is not something we understand, check the status code to determine if it is a success
+		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+			return nil
+		}
+		return errors.Join(err, &smsapi.SendError{
+			DumpedResponse: dumpedResponse,
+		})
 	}
+
+	return handleResponse(responseBody, dumpedResponse)
 }
 
-func NewSMSDenoHook(lf *log.Factory, denoEndpoint config.DenoEndpoint, smsCfg *config.CustomSMSProviderConfig) *SMSDenoHook {
+func NewSMSDenoHookForTest(lf *log.Factory, denoEndpoint config.DenoEndpoint, smsCfg *config.CustomSMSProviderConfig) *SMSDenoHook {
 	timeout := NewSMSHookTimeout(smsCfg)
 	logger := hook.NewLogger(lf)
 	client := NewHookDenoClient(denoEndpoint, logger, timeout)
+	// DenoHook is not needed because it can only be used for Test()
 	return &SMSDenoHook{
 		Client: client,
 	}
 }
 
+type DenoHook interface {
+	RunSync(ctx context.Context, client hook.DenoClient, u *url.URL, input interface{}) (out interface{}, err error)
+	SupportURL(u *url.URL) bool
+}
+
 type SMSDenoHook struct {
-	hook.DenoHook
+	DenoHook
 	Client HookDenoClient
 }
 
-func (d *SMSDenoHook) Call(ctx context.Context, u *url.URL, payload SendOptions) ([]byte, error) {
+func (d *SMSDenoHook) Call(ctx context.Context, u *url.URL, payload SendOptions) error {
 	anything, err := d.RunSync(ctx, d.Client, u, payload)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	jsonText, err := json.Marshal(anything)
+	return d.handleOutput(anything)
+}
+
+func (d *SMSDenoHook) Test(ctx context.Context, script string, payload SendOptions) error {
+	anything, err := d.Client.Run(ctx, script, payload)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	return jsonText, nil
+	return d.handleOutput(anything)
 }
 
-func (d *SMSDenoHook) Test(ctx context.Context, script string, payload SendOptions) error {
-	_, err := d.Client.Run(ctx, script, payload)
+func (d *SMSDenoHook) handleOutput(output interface{}) error {
+	if output == nil {
+		// This is a null, but we should still consider it is a success for backward compatibility.
+		return nil
+	}
+
+	jsonText, err := json.Marshal(output)
 	if err != nil {
 		return err
 	}
 
-	return nil
+	responseBody, err := ParseResponseBody(jsonText)
```

Comment:
If the existing SMS deno hook return `{"code": 1}`, the json parsing will fail, but we should still treat this as success. Please add a test case to show this.

Reply:
Currently, sms deno hook is not allowed to return anything. So existing deno hooks should return `null`.

Reply:
What do you mean by not allowed? I think the hook can return anything they want, given that they write enough type annotation to bypass type checking.

Reply:
You are right that the hook can return anything if the function return type is set to any. However I think existing hooks are configured by us so there should be no hook written using incorrect return type? In fact I am not sure if there is anyone using the deno sms hook after we have our sms gateway deployed?